Log a warning when assertions are incorrectly nested and errors are raised

ghiculescu opened the pull request:
Follow up to #37313

- Adds regression tests
- Logs a warning in cases where assertions are nested in a way that's likely to be confusing

Note that my original implementation basically reverted https://github.com/rails/rails/pull/37313 which is what some of the PR comments refer to.

ghiculescu committed:
Log a warning when assertions are incorrectly nested and errors are raised

Follow up to https://github.com/rails/rails/pull/37313

- Adds regression tests
- Logs a warning in cases where assertions are nested in a way that's likely to be confusing

p8 commented:
@ghiculescu I don't think it's a good idea to nest assertions , especially adding an assertion in an `assert_raise` block.
I'd probably write that failing test like:
```ruby
    email = nil
    assert_raises(ArgumentError) do
      email = MyMailer.send_email("foo")
    end
    assert_emails(0) { email.deliver_now }
```
Although `email` would be `nil` on the last line.

ghiculescu commented:
@p8 the call to deliver_now raises. I think nested assertions are a fairly common practice, I’ve never seen them warned against.

p8 commented:
@ghiculescu ah my bad.
Still I'd probably create a separate test for each assertion.
I'd also move the assignment to `email` out of the block to make it clear that doesn't raise:

```ruby
  def test_it_raises_an_error
    email = MyMailer.send_email("foo")
    assert_raises(ArgumentError) do
       email.deliver_now
    end
  end
  
  def test_it_sends_no_mail
      email = MyMailer.send_email("foo")
      assert_emails(0) { email.deliver_now }
    end
  end
``` 

Because you might accidentally do:
```
    assert_raises(ArgumentError) do
        assert_emails(:bad_arg) { email.deliver_now }
    end
```

In rspec you can chaing assertions and do something like:

```ruby
  it "does not send mail" do
    email = MyMailer.send_email("foo")
    expect { email.deliver_now }
      .to_raise(ArgumentError)
      .and not_change { BaseMailer.count }
  end
```

ghiculescu commented:
Yeah, that works. I ended up switching the order of assertion nesting, which also got the test to pass. But I’m raising this issue because I don’t think the order of assertion raising should influence the test outcome, or if it does, it should be way more obvious (to people who don’t know rails internals / read rails PRs) how to fix it.

eugeneius commented:
Since `assert_nothing_raised` is now a no-op again, this logically reverts https://github.com/rails/rails/pull/37313, and reintroduces the problem it solved.

> I don’t think the order of assertion raising should influence the test outcome

I assume you meant "the order of assertion nesting" here - that was also the motivation for https://github.com/rails/rails/pull/37313.

It seems like there's a tradeoff here between two similar failure modes:

 - Implementation is wrong, test passes (fixed by https://github.com/rails/rails/pull/37313).
 - Implementation is correct, test fails (fixed by this PR).

I personally prefer the current state of affairs; false negatives are much worse than false positives.

matthewd commented:
> Since assert_nothing_raised is now a no-op again, this logically reverts #37313, and reintroduces the problem it solved.

Do we need some tests to protect that behaviour?

How does that solution compare to, moving all block-assertion-methods' post-yield checks into `ensure`?

Or, for something closer to this change: is it reasonable for `assert_nothing_raised` to re-raise assertion failures unchanged, while wrapping all other exceptions with a failure? (Would that solve the problem here? Would it solve the original problem?)

ghiculescu commented:
Yeah, my plan is to add tests and reconsider #3731e before taking this out of draft mode. Sorry if I implied this was ready now! @matthewd’s idea is where I expect to end up, but with tests so we can be confident.

casperisfine commented:
I'm a hard no on this.

Consider this test on 6.0:

```ruby
assert_raises(ArgumentError) do
  email = MyMailer.send_email(foo)
  assert_emails(42) { email.deliver_now }
end
```

It would pass.

From the beginning your `assert_emails` was a total noop and wasn't asserting anything, the change in 6.1 fixed that.

Also note that this behavior was also changed in minitest itself.

ghiculescu commented:
This is still a work in progress, but I've added a warning that gets triggered if an internal assertion added by https://github.com/rails/rails/pull/37313 swallows an error. I've also added some regression tests for https://github.com/rails/rails/pull/37313. Keen for feedback on the direction it's taking.

ghiculescu commented on activesupport/test/test_case_test.rb:
```diff
@@ -350,6 +350,85 @@ def test_assert_no_changes_with_long_string_wont_output_everything
+  private
+    def run_test_that_should_pass_and_log_a_warning
+      assert_raises(Minitest::UnexpectedError) do # this assertion passes, but it's unlikely to be how anyone writes a test
+        assert_no_changes -> { 1 } do # this assertion doesn't run. the error below is caught and the error logged.
+          raise ArgumentError.new
+        end
+      end
+    end
+
+    def run_test_that_should_fail_confusingly
+      assert_raises(ArgumentError) do # this assertion fails (confusingly) because it catches a Minitest::UnexpectedError.
+        assert_no_changes -> { 1 } do # this assertion doesn't run. the error below is caught and the error logged.
```
```suggestion
        assert_no_changes -> { 1 } do # this assertion doesn't run. the error below is caught and the warning logged.
          raise ArgumentError.new
        end
      end
    end

    def run_test_that_should_fail_confusingly
      assert_raises(ArgumentError) do # this assertion fails (confusingly) because it catches a Minitest::UnexpectedError.
        assert_no_changes -> { 1 } do # this assertion doesn't run. the error below is caught and the warning logged.
```

byroot commented on activesupport/lib/active_support/testing/assertions.rb:
```diff
@@ -244,6 +244,22 @@ def assert_no_changes(expression, message = nil, from: UNTRACKED, &block)
+        rescue Minitest::UnexpectedError => e
+          if tagged_logger && tagged_logger.warn?
+            warning = <<~MSG
+              #{self.class} - #{name}: assert_nothing_raised called but #{e.error.class} raised.
```
`assert_nothing_raised called` isn't ideal. In many cases you dont' call it yourself, you call e.g. `assert_difference {}`, which calls `assert_nothing_raised`. Maybe the caller name could be passed in argument or something.

ghiculescu replied:
Yep, will make the warning more clear.

ghiculescu replied:
I think the simplest way is just to remove the mention of `assert_nothing_raised` and keep the rest. So I did that.

byroot commented on activesupport/lib/active_support/testing/assertions.rb:
```diff
@@ -244,6 +244,22 @@ def assert_no_changes(expression, message = nil, from: UNTRACKED, &block)
+            tagged_logger.warn warning
+          end
+
+          raise(e)
```
Just `raise` (without arguments) avoid changing the backtrace.

byroot commented on activesupport/lib/active_support/testing/assertions.rb:
```diff
@@ -244,6 +244,22 @@ def assert_no_changes(expression, message = nil, from: UNTRACKED, &block)
+          if tagged_logger && tagged_logger.warn?
+            warning = <<~MSG
+              #{self.class} - #{name}: #{e.error.class} raised.
+              Assertions nested *inside* `assert_raises` are ignored, this could cause tests to pass incorrectly or fail confusingly.
```
Again I'm not good at copy, but `Assertions nested *inside* `assert_raises` are ignored` is weird to me, because they aren't ignored anymore.

Maybe the message should instead say that if the exception was expected, it should be with an `assert_raises` inside the other block assertion?

ghiculescu replied:
Yeah I'm really struggling getting the wording right here. Thanks for challenging me on it. Pushed a third attempt.

byroot reviewed (approved):
Makes much more sense to me now!

ghiculescu commented:
thanks for your help + patience on this one @byroot

pirj commented on activesupport/lib/active_support/testing/assertions.rb:
```diff
@@ -244,6 +244,22 @@ def assert_no_changes(expression, message = nil, from: UNTRACKED, &block)
+        def _assert_nothing_raised_or_warn(assertion, &block)
+          assert_nothing_raised(&block)
+        rescue Minitest::UnexpectedError => e
+          if tagged_logger && tagged_logger.warn?
```
Won't this result in a
```
got #<NameError: undefined local variable or method `tagged_logger'
```
if you don't explicitly set `self.tagged_logger =`, @ghiculescu ?
WDYT of `defined?(tagged_logger) && tagged_logger&.warn?` instead?

Where is `tagged_logger` normally assigned to be accessible in this context?

ghiculescu replied:
I think I copied this from https://github.com/rails/rails/blob/18707ab17fa492eb25ad2e8f9818a320dc20b823/activesupport/lib/active_support/testing/tagged_logging.rb#L11

That module is included here https://github.com/rails/rails/blob/18707ab17fa492eb25ad2e8f9818a320dc20b823/activesupport/lib/active_support/test_case.rb#L129

I can change to a `defined?` check if needed, but curious how you have it set up that it's breaking for you?

rafaelfranca replied:
We should not change it. Rails is assuming correctly that this method exists because it does in all `ActiveSupport::TestCase`. This problem only happens in RSpec, so RSpec should be responsible to makes sure that method exist.
